Fix Inductor CSE Across Separate Reductions

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119410

We were CSE'ing a load across two separate reduction loop bodies. This is because we were examining an indirect indexing that did not have an explicit rindex in its load. I've commented with more details and other potentials on the fix.

Tried using minifier unsuccessfully and hand minified some but could do more..


Fix for https://github.com/pytorch/pytorch/issues/119327

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

eellison committed:
Fix Inductor CSE Across Separate Reductions

[ghstack-poisoned]

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1904,7 +1913,7 @@ def load(self, name: str, index: sympy.Expr):
             load_buffer.writeline(advance_block_ptr)
 
-        if not self.inside_reduction or not has_rindex:
+        if not self.inside_reduction or not indexing.has_rmask():
```
```
        tmp0 = tl.load(in_ptr0 + (r1), rmask, eviction_policy='evict_last', other=0.0)
        tmp1 = tmp0 + 4191
        tmp2 = tmp0 < 0
        tmp3 = tl.where(tmp2, tmp1, tmp0)
        tl.device_assert(((0 <= tmp3) & (tmp3 < 4191)) | ~rmask, "index out of bounds: 0 <= tmp3 < 4191")
        tmp4 = tl.load(in_ptr1 + (tmp3 + (4191*x0)), rmask, eviction_policy='evict_last')
 ```
 
 We were incorrectly cse-ing `tmp4` because it's load did not have an rindex. I've put the condition here back to what it was prior to [this pr](https://github.com/pytorch/pytorch/pull/116079).
 
 We could also make the `has_rindex` propagate correctly from tmp0 -> tmp4. Or ban indirect indexing like [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L1887-L1891).

shunting314 replied:
Is the rmask for loading in_ptr1 here get propagated from the rmask used when loading from in_ptr0?

peterbell10 replied:
Yes, `TritonCSEVariable` propagates mask information to intermediate variables so that indirect indexes have all the required masks.

eellison committed:
Update on "Fix Inductor CSE Across Separate Reductions"


We were CSE'ing a load across two separate reduction loop bodies. This is because we were examining an indirect indexing that did not have an explicit rindex in its load. I've commented with more details and other potentials on the fix.

Tried using minifier unsuccessfully and hand minified some but could do more..


Fix for https://github.com/pytorch/pytorch/issues/119327

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 reviewed (approved):
maybe find a test easier to manage and understand

eellison committed:
Update on "Fix Inductor CSE Across Separate Reductions"


We were CSE'ing a load across two separate reduction loop bodies. This is because we were examining an indirect indexing that did not have an explicit rindex in its load. I've commented with more details and other potentials on the fix.

Tried using minifier unsuccessfully and hand minified some but could do more..


Fix for https://github.com/pytorch/pytorch/issues/119327

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Fix Inductor CSE Across Separate Reductions"


We were CSE'ing a load across two separate reduction loop bodies. This is because we were examining an indirect indexing that did not have an explicit rindex in its load. I've commented with more details and other potentials on the fix.

Tried using minifier unsuccessfully and hand minified some but could do more..


Fix for https://github.com/pytorch/pytorch/issues/119327

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Fix Inductor CSE Across Separate Reductions"


We were CSE'ing a load across two separate reduction loop bodies. This is because we were examining an indirect indexing that did not have an explicit rindex in its load. I've commented with more details and other potentials on the fix.

Tried using minifier unsuccessfully and hand minified some but could do more..


Fix for https://github.com/pytorch/pytorch/issues/119327

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Fix Inductor CSE Across Separate Reductions"


We were CSE'ing a load across two separate reduction loop bodies. This is because we were examining an indirect indexing that did not have an explicit rindex in its load. I've commented with more details and other potentials on the fix.

Tried using minifier unsuccessfully and hand minified some but could do more..


Fix for https://github.com/pytorch/pytorch/issues/119327

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Fix Inductor CSE Across Separate Reductions"


We were CSE'ing a load across two separate reduction loop bodies. This is because we were examining an indirect indexing that did not have an explicit rindex in its load. I've commented with more details and other potentials on the fix.

Tried using minifier unsuccessfully and hand minified some but could do more..


Fix for https://github.com/pytorch/pytorch/issues/119327

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update on "Fix Inductor CSE Across Separate Reductions"


We were CSE'ing a load across two separate reduction loop bodies. This is because we were examining an indirect indexing that did not have an explicit rindex in its load. I've commented with more details and other potentials on the fix.

Tried using minifier unsuccessfully and hand minified some but could do more..


Fix for https://github.com/pytorch/pytorch/issues/119327

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1948,7 +1957,7 @@ def load(self, name: str, index: sympy.Expr):
             load_buffer.writeline(advance_block_ptr)
 
-        if not self.inside_reduction or not has_rindex:
+        if not self.inside_reduction or (not indexing.has_rmask() and not has_rindex):
```
shouldn't this condition be 
```
if not self.inside_reduction or load_buffer == self.body:
```
? That is, a variable is outside a loop if there was no loop to start with or we were able to lift it to he body of the function.
